class_name Player
extends CharacterBody3D

# Constantes pour la vitesse et la rotation
const SPEED: float = 2.0
const ROTATION_SPEED: float = 10.0

# R√©f√©rences aux n≈ìuds de la sc√®ne
var camera_point: Node3D

@onready var camera_point_1: Node3D = %camera_point1
@onready var camera_point_2: Node3D = %camera_point2
@onready var camera_point_3: Node3D = %camera_point3

@onready var animation_player: AnimationPlayer = %AnimationPlayer

@onready var visuals: Node3D = $Visuals

# R√©cup√©ration de la gravit√© du projet
var gravity: float = ProjectSettings.get_setting("physics/3d/default_gravity")

# √âtats du joueur
var walking: bool = false
var can_interact_object: bool = false

# Inventaire
@export var inventory: Inventory

# Objets √† proximit√©
var nearby_object: Objects = null
var curent_basin: Basin = null
var curent_source: Source = null
var node_object: Node3D = null

func _ready():
	set_camera_point_1()
	GameManager.set_player(self)
	animation_player.play("idle2")
	# Ajout des transitions fluides entre les animations
	animation_player.set_blend_time("idle2", "walk2", 0.3)
	animation_player.set_blend_time("walk2", "idle2", 0.3)
	
	animation_player.set_blend_time("idle2", "cooking", 0.3)
	animation_player.set_blend_time("cooking", "idle2", 0.3)
	
	animation_player.set_blend_time("cooking", "walk2", 0.3)
	animation_player.set_blend_time("walk2", "cooking", 0.3)
	

func _physics_process(delta: float):
	_apply_gravity(delta)
	_handle_movement(delta)
	_handle_interaction()

	move_and_slide()

func set_camera_point_1():
	camera_point = camera_point_1

func set_camera_point_2():
	camera_point = camera_point_2
	
func set_camera_point_3():
	camera_point = camera_point_3

# Applique la gravit√© si le joueur n'est pas au sol
func _apply_gravity(delta: float):
	if not is_on_floor():
		velocity.y -= gravity * delta

# G√®re les d√©placements et l'animation du joueur
func _handle_movement(delta: float):
	var input_dir: Vector2 = Input.get_vector("left", "right", "forward", "backward")
	var direction: Vector3 = (transform.basis * Vector3(input_dir.x, 0, input_dir.y)).normalized()

	if direction != Vector3.ZERO:
		velocity.x = direction.x * SPEED
		velocity.z = direction.z * SPEED

		var iso_direction: Vector3 = Vector3(
			-direction.x + direction.z,
			0,
			-direction.x - direction.z
		).normalized()

		# Rotation fluide du personnage
		var target_angle: float = atan2(iso_direction.x, iso_direction.z)
		visuals.rotation.y = lerp_angle(visuals.rotation.y, target_angle, ROTATION_SPEED * delta)

		# Lancement de l'animation de marche si n√©cessaire
		if not walking:
			walking = true
			animation_player.play("walk2")
	else:
		velocity.x = move_toward(velocity.x, 0, SPEED)
		velocity.z = move_toward(velocity.z, 0, SPEED)
		
		# Retour √† l'animation d'idle si le joueur s'arr√™te
		if walking:
			walking = false
			animation_player.play("idle2")

# G√®re les interactions avec les objets
func _handle_interaction():
	#if Input.is_action_just_pressed("interact") and nearby_object:
		#print("Interaction avec : " + nearby_object.name)
#
		#if GameInteraction.pick and not GameInteraction.drop:
			#pick_object(nearby_object)
		#elif GameInteraction.drop and not GameInteraction.pick and inventory._content.size() > 0:
			#if nearby_object is Recipe:
				#place_object(nearby_object)
			#else:
				#print("Impossible de placer cet objet ici.")
#
	## Permet de d√©poser un objet m√™me s'il n'y a rien √† proximit√©
	#elif Input.is_action_just_pressed("interact") and GameInteraction.drop and not GameInteraction.pick and inventory._content.size() > 0 and !can_interact_object:
		#drop_all_objects()
	pass

# Ramasse un objet et l'ajoute √† l'inventaire
func pick_object(item: Item):
	print("Objet ramass√©: ", item.name)
	inventory.add_item(item)

# D√©pose un objet depuis l'inventaire
func drop_all_objects():
	if inventory._content.is_empty():
		print("‚ö†Ô∏è Aucun objet √† d√©poser.")
		return
	
	for item in inventory.get_items():
		print("üóëÔ∏è Objet d√©pos√© :", item.name)
		#inventory.remove_item(item, 1)
	
	print("‚úÖ Tous les objets ont √©t√© d√©pos√©s.")


# Place un objet dans un emplacement sp√©cifique
func place_object(recipe: Recipe):
	# On r√©cup√®re la liste des ingr√©dients que le joueur poss√®de
	var ingredients_to_place = []
	for item in inventory._content:
		if item in recipe.ingredients and item not in recipe.placed_ingredients:
			ingredients_to_place.append(item)
			print("Objet plac√© dans un : ", recipe.name)

	# Ajout des ingr√©dients √† la recette
	for item in ingredients_to_place:
		recipe.add_ingredient(item)
		inventory.remove_item(item)
		# Affichage du pourcentage de progression
		var completion = recipe.get_completion_percentage()
		print("Progression de la recette :", completion, "%")
